doc/compiler/optimizer/ValuePropagation.md

```diff
@@ -0,0 +1,195 @@
+
+A `merge` of two constraints will weaken the constraint. For example consider the following program:
+
+```
+if (...) {
+    if (x != 10) 
+        return;
+    ...
+} else {
+    if (x != 5)
+       return;
+    ...
+}
+
+...
+```
+
+After analyzing both the outer and inner `if` and the `else` blocks the analysis needs to continue with the merge point of the control flow. At that point we must `merge` the set of constraints we have determined within the outer `if` and the `else` blocks. In the above example the value number of `x` following the merge point would have a constraint which merges the constraints from within the previous blocks, i.e. a discrete range constraint specifying that the value number of `x` can be in ranges `[10, 10]` or `[5, 5]`, since the only way to reach the merge point is if we did not execute the `return` statement.
+
+Note that we typically don't represent discrete values, but rather ranges of discrete values, which are easier to merge with one another.
+
+In summary, merging constraints only makes the information known at the time worse, or at the same level as before.
+
+## Examples of Constraints
+
+### Range Constraints
+
+These constraints represent ranges of values a particular value number can take at a given point in the program. As we saw previously these ranges can be discrete values, or true continuous (potentially discrete) ranges. Range constraints can be used to eliminate control flow or propagate particular values to different points in the program. There are many range constraint classes for tracking different types, such as integers and longs.
+
+### Type Constraints
+
+For typed languages it is often beneficial to flow type constraints. Knowing type information about a particular value number at a given point in the program can allow us to devirtualize calls or eliminate type check and exception edges.
+
+There are three different constraint classes which flow this information in VP, they are the [fixed class](https://github.com/eclipse/omr/blob/25f479d2b671c2a0a03156cc0a48024a46d2e3c2/compiler/optimizer/VPConstraint.hpp#L655-L674) constraint, [resolved class](https://github.com/eclipse/omr/blob/25f479d2b671c2a0a03156cc0a48024a46d2e3c2/compiler/optimizer/VPConstraint.hpp#L628-L653) constraint, and [unresolved class](https://github.com/eclipse/omr/blob/25f479d2b671c2a0a03156cc0a48024a46d2e3c2/compiler/optimizer/VPConstraint.hpp#L706-L736) constraints.
+
+Fixed class constraints simply express that a value number has a specific, fixed, type `T`. We can know this information perhaps because we have seen the allocation of the value number expression or perhaps we have seen an explicit cast into type `T`. Knowing the exact type of a value number expression is the most useful.
+
```

Comment:
An explicit cast into type `T` is not usually enough to generate a fixed class constraint. The type that we cast to must be `final` in order for the constraint to be a fixed one, otherwise it just stays a resolved class constraint.

Reply:
Fixed in [force-push](https://github.com/eclipse/omr/compare/5c1aa757ca3e116307e60404b9d3b0743e2ce03f..127bdeabc6860a9f211446b05d6210a084512605).